Add docs about translation handling (#6405)
woocommerce/woocommerce-blocks: docs/translations/translation-loading.md

# Translation files and loading

## Translation files

### POT (Portable Object Template) file

> POT stands for `Portable Object Template` and contains all original strings, in English. It can be created manually using [WP-CLI](https://wp-cli.org/) or [Poedit](https://poedit.net/). As the WooCOmmerce Blocks plugin is hosted on WordPress.org, we don't need to manually create the POT file. GlotPress automatically generates that file in the background, once we release a new version of the plugin.

Reviewer: ```suggestion
POT stands for `Portable Object Template` and contains all original strings, in English. It can be created manually using [WP-CLI](https://wp-cli.org/) or [Poedit](https://poedit.net/). As the WooCommerce Blocks plugin is hosted on WordPress.org, we don't need to manually create the POT file. GlotPress automatically generates that file in the background, once we release a new version of the plugin.
```

```po
# Translation of Plugins - WooCommerce Blocks - Stable (latest release) in German
# This file is distributed under the same license as the Plugins - WooCommerce Blocks - Stable (latest release) package.
msgid ""
msgstr ""
"PO-Revision-Date: 2022-05-22 10:58:25+0000\n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: 8bit\n"
"Plural-Forms: nplurals=2; plural=n != 1;\n"
"X-Generator: GlotPress/4.0.0-alpha.1\n"
"Language: de\n"
"Project-Id-Version: Plugins - WooCommerce Blocks - Stable (latest release)\n"

#: assets/js/blocks/featured-product/block.json
#: build/featured-product/block.json
msgctxt "block description"
msgid "Visually highlight a product or variation and encourage prompt action."
msgstr "Ein Produkt oder eine Variante visuell hervorheben und zum sofortigen Handeln auffordern."

#: assets/js/blocks/featured-product/block.json
#: build/featured-product/block.json
msgctxt "block title"
msgid "Featured Product"
msgstr "Hervorgehobenes Produkt"

[...]
```

> See also <https://developer.wordpress.org/plugins/internationalization/localization/#po-portable-object-files>.

Reviewer 2: Not sure if this is worth noting, but if we create the `po` and `mo` files manually from the pot file (es explained above), then putting those files in `plugin-root-dir/languages/` won't work because we don't load the text domain using `load_plugin_textdomain()`.

This can be confused because the JSON translations that exist in `plugin-root-dir/languages/`, on the opposite, are loaded and have a higher priority than ones in `wp-content/languages/plugins/`.

Author: Based on https://developer.wordpress.org/plugins/internationalization/how-to-internationalize-your-plugin/#plugins-on-wordpress-org, I understood that since WordPress 4.6 `load_plugin_textdomain()` is no longer required if the plugin is hosted on WordPress.org and the plugin slug and the text domain are equal. Is that information incorrect?

Reviewer 2: > since WordPress 4.6 `load_plugin_textdomain()` is no longer required if the plugin is hosted on WordPress.org and the plugin slug and the text domain are equal. Is that information incorrect?

@nielslange Yes. That's correct.

I was talking about the case when developers manually create `pot`, then `po` and `mo` files locally. When working with translation issues, I created PO and MO files manually and put them in the `plugin-dir/languages`, which won't work because we don't load po/mo files from the `plugin-dir/languages`.

It may be only I encountered that issue so it's okay to keep this file as it is now.

Reviewer 2: Just want to clarify my point:

Without `load_plugin_textdomain` in the plugin, only PO and MO files from `wp-content/languages/plugins` will be loaded. Putting those files in `plugin-dir/languages` won't work.

This is never an issue if translations are downloaded automatically. But it may be an issue if developers create them manually.

### MO (Machine Object) file

MO stands for `Machine Object`, contains both the original and the translated strings and is based on the PO file. As the POT and the PO files, the MO file can be created manually, but this is not necessary for the WooCommerce Blocks plugin.

> The MO file is only machine-readable and named `woo-gutenberg-products-block-{LANGUAGE-CODE}.mo`, e.g. `woo-gutenberg-products-block-de_DE.mo` It can be found in the `/wp-content/languages/plugin` folder. As the MO file is machine-readable only, it cannot be viewed and the MO file only handeles translations within PHP files.

Reviewer: ```suggestion
The MO file is only machine-readable and named `woo-gutenberg-products-block-{LANGUAGE-CODE}.mo`, e.g. `woo-gutenberg-products-block-de_DE.mo` It can be found in the `/wp-content/languages/plugin` folder. As the MO file is machine-readable only, it cannot be viewed and the MO file only handles translations within PHP files.
```

See also https://developer.wordpress.org/plugins/internationalization/localization/#mo-machine-object-files.

### JSON files

> As mentioned before, the MO file only handle translations within PHP files. Translations within JavaScript and TypeScript files are handeled by JSON files. These JSON files contain both the original and the translated strings and are also based on the PO file. Similar to the MO file, the JSON files can be manually created, which is not needed in this case.

Reviewer: ```suggestion
As mentioned before, the MO file only handle translations within PHP files. Translations within JavaScript and TypeScript files are handled by JSON files. These JSON files contain both the original and the translated strings and are also based on the PO file. Similar to the MO file, the JSON files can be manually created, which is not needed in this case.
```

The JSON files are human-readable and named `woo-gutenberg-products-block-{LANGUAGE-CODE}-{md5(FILE-PATH)}.json`, e.g. `woo-gutenberg-products-block-de_DE-0a066f0c536e17452f6345c5d072335b.json`. They can be found in the `/wp-content/languages/plugin` folder. The JSON files look like this:

```json
{
	"translation-revision-date": "2022-05-22 10:58:25+0000",
	"generator": "GlotPress/4.0.0-alpha.1",
	"domain": "messages",
	"locale_data": {
		"messages": {
			"": {
				"domain": "messages",
				"plural-forms": "nplurals=2; plural=n != 1;",
				"lang": "de"
			},
			"Price between %1$s and %2$s": [ "Preis zwischen %1$s und %2$s" ],
			"Discounted price:": [ "Reduzierter Preis:" ],
			"Previous price:": [ "Vorheriger Preis:" ]
		}
	},
	"comment": { "reference": "build/product-price-frontend.js" }
}
```

## Loading translations

> Loading translations for PHP files works different from load translations for JS/TS files.

Reviewer: ```suggestion
Loading translations for PHP files works differently to loading translations for JS/TS files.
```

### Loading translations for PHP files

As mentioned in [Translation Basics](translation-basics.md), loading the translations for PHP files does not regquire any extra code, as long as the plugin is hosted on WordPress.org and does not support WordPress versions prior to 4.6.

### Loading translations for TJS/TS files

> To load JS/TS translations, the function `wp_set_script_translations()` needs to be called. Currently, this function is part of the following code block in `/src/Assets/Api.php`:

Reviewer: ```suggestion
To load JS/TS translations, you need to execute the `wp_set_script_translations()` function. Currently, this function is part of the following code block in `/src/Assets/Api.php`:
```

Passive -> active voice :)

              printf( esc_html__( 'Script with handle %s had a dependency on itself which has been removed. This is an indicator that your JS code has a circular dependency that can cause bugs.', 'woo-gutenberg-products-block' ), esc_html( $handle ) );
              echo '</p></div>';
          }
        );
    } else {
      throw new Exception( sprintf( 'Script with handle %s had a dependency on itself. This is an indicator that your JS code has a circular dependency that can cause bugs.', $handle ) );
    }
  }

  /**
   * Filters the list of script dependencies.
   *
   * @param array $dependencies The list of script dependencies.
   * @param string $handle The script's handle.
   * @return array
   */
  $script_dependencies = apply_filters( 'woocommerce_blocks_register_script_dependencies', $script_data['dependencies'], $handle );

  wp_register_script( $handle, $script_data['src'], $script_dependencies, $script_data['version'], true );

  if ( $has_i18n && function_exists( 'wp_set_script_translations' ) ) {
    wp_set_script_translations( $handle, 'woo-gutenberg-products-block', $this->package->get_path( 'languages' ) );
  }
}
```

## Loading fallback translations

> By default, the WooCommerce Blocks plugin tries to load the translation from https://translate.wordpress.org/projects/wp-plugins/woo-gutenberg-products-block/. In case a translation cannot be loaded, the plugin then tries to load the correspnding translation from https://translate.wordpress.org/projects/wp-plugins/woocommerce/.

Reviewer: ```suggestion
By default, the WooCommerce Blocks plugin tries to load the translation from https://translate.wordpress.org/projects/wp-plugins/woo-gutenberg-products-block/. If a translation cannot be loaded, the plugin tries to load the corresponding translation from https://translate.wordpress.org/projects/wp-plugins/woocommerce/.
```

		wp_register_script( $handle_filename, '', array( 'wp-i18n' ), false, true );
		wp_enqueue_script( $handle_filename );
		wp_add_inline_script(
			$handle_filename,
			$output
		);
	} else {
		// If we have already printed scripts into the page, there is a chance that
		// scripts have finished being printed. That means that if we enqueued them here,
		// they would never be printed. Instead of enqueuing, then, let's print directly
		// the script tag.
		printf( "<script type='text/javascript'>\n%s\n</script>\n", $output ); // phpcs:ignore
	}

	// Finally, short circuit the pre_load_script_translations hook by returning
	// the translation JSON from the feature plugin, if it exists so this hook
	// does not run again for the current handle.
	$path_md5     = md5( 'build/' . $handle_filename );
	$json_file    = $lang_dir . '/' . $domain . '-' . $locale . '-' . $path_md5 . '.json';
	$translations = is_file( $json_file ) && is_readable( $json_file ) ? file_get_contents( $json_file ) : false; // phpcs:ignore

	if ( $translations ) {
		return $translations;
	}

	// Return valid empty Jed locale.
	return '{ "locale_data": { "messages": { "": {} } } }';
}

> add_filter( 'pre_load_script_translations', 'woocommerce_blocks_get_i18n_data_json', 10, 4 );

Reviewer 2: This filter and corresponding callback were refactored in #6478. Do you think we should update this example to prevent confusion?

Author: I've just updated that code example, as suggested by you.
